### ISSUE TITLE
BulkAll onCompleted is never called in v2.5.3+

### ISSUE THREAD

**Issue by Friss** [elastic/elasticsearch-net#2731]
Looks like the fix for #2644 (https://github.com/elastic/elasticsearch-net/pull/2666) broke the `onComplete` functionality. Reverting to 2.5.2 fixes the issue. 

**NEST/Elasticsearch.Net version**: 2.5.3+

**Elasticsearch version**: 2.4.1

`BulkAllObserver` `onComplete` is never called 

**Steps to reproduce**:
 1. Have a bunch of documents to index
 2. Run `BulkAll` with a `BulkAllObserver` subscribed with an `onComplete` function

```
var waitComplete = new CountdownEvent(1);
var waitOneBulk = new CountdownEvent(1);

var observableBulkAll = client.BulkAll(documents, f => f
  .MaxDegreeOfParallelism(2)
  .BackOffTime(TimeSpan.FromSeconds(15))
  .BackOffRetries(3)
  .Size(Constants.INDEX_PARTITION_SIZE)
  .Index("documents")
);

var bulkAllObserver = new BulkAllObserver(
  onError: (e) => { throw e; },
  onCompleted: () => {
    // Never gets called
    waitComplete.Signal();
  },
  onNext: (b) => {
    try {
      waitOneBulk.Signal();
    } catch (Exception) { }
  }
);

observableBulkAll.Subscribe(bulkAllObserver);
waitOneBulk.Wait();
waitComplete.Wait();
```

**Comment by Mpdreamz** [comment]
I can not reproduce this:

This is from my linqpad session (where .Dump() is available).

```csharp
var waitComplete = new CountdownEvent(1);
var waitOneBulk = new CountdownEvent(1);

var host = new Uri("http://localhost:9200");
var settings = new ConnectionSettings(new SingleNodeConnectionPool(host));
var client = new ElasticClient(settings);

client.DeleteIndex("documents");

var documents = Enumerable.Range(0, 1000)
	.Select(i => new { Id = i });

var observableBulkAll = client.BulkAll(documents, f => f
  .MaxDegreeOfParallelism(2)
  .BackOffTime(TimeSpan.FromSeconds(15))
  .BackOffRetries(3)
  .Size(100)
  .Index("documents")
  .Type("doc")
  .RefreshOnCompleted()
);

var bulkAllObserver = new BulkAllObserver(
  onError: (e) => { throw e; },
  onCompleted: () =>
  {
	  // Never gets called
	  waitComplete.Signal();
  },
  onNext: (b) =>
  {
	  try
	  {
		  waitOneBulk.Signal();
	  }
	  catch (Exception) { }
  }
);

observableBulkAll.Subscribe(bulkAllObserver);
waitOneBulk.Wait();
waitComplete.Wait();

//can call this due to RefreshOnCompleted on observer

var r = client.Count<object>(c => c.Index("documents").Type("doc"));
var rr = client.RootNodeInfo();

rr.Version.Number.Dump("elasticsearch version");
r.Count.Dump();
```

I can see the breakpoint being hit:

![image](https://cloud.githubusercontent.com/assets/245275/25429056/fe3de09e-2a77-11e7-9f3b-6ea77f27d2b3.png)

And both the  `Wait()` gets signaled

What I think happens is  that `onError` is triggered but that `throw e` is thrown on a different thread and not bubbling out on the thread waiting.  

The semantics here have changed following the fix in #2644.

To help you with this we ship with ` Wait(TimeSpan timeout, Func onNext = null)` extensions on the observable we ship which will make sure to rethrow as you'd expect them too on the waiting thread.

**Comment by Friss** [comment]
Hmm I'll double check in my code but the only thing that changed was upgrading nest and the deps to the latest version on the 2.x branch. 

And I don't think it would make much of a difference in isolation but what I am really doing is calling BulkAll multiple times and storing the CountdownEvent in a list to ensure all have been signaled before exiting.
